src/v/serde/protobuf/parser.cc

```diff
@@ -298,14 +300,73 @@ class parser {
         }
     }
 
-    template<typename T>
-    void update_field(int32_t field_number, T&& value) {
+    void
+    update_field(int32_t field_number, std::unique_ptr<parsed::message> value) {
         auto* field = current_->descriptor->FindFieldByNumber(field_number);
         if (field == nullptr) {
             throw std::runtime_error(
               fmt::format("unknown field number: {}", field_number));
+        }
+        if (field->is_map()) {
+            convert_to_map_and_update(*field, std::move(value));
         } else {
-            update_field(*field, std::forward<T>(value));
+            update_field(*field, std::move(value));
+        }
+    }
+
+    void convert_to_map_and_update(
+      const pb::FieldDescriptor& field,
+      std::unique_ptr<parsed::message> value) {
+        const auto* key_descriptor = field.message_type()->map_key();
+        parsed::map::key key = std::monostate{};
+        auto it = value->fields.find(key_descriptor->number());
+        if (it != value->fields.end()) {
+            key = ss::visit(
+              it->second,
+              [](double val) -> parsed::map::key { return val; },
+              [](float val) -> parsed::map::key { return val; },
+              [](int32_t val) -> parsed::map::key { return val; },
+              [](int64_t val) -> parsed::map::key { return val; },
+              [](uint32_t val) -> parsed::map::key { return val; },
+              [](uint64_t val) -> parsed::map::key { return val; },
+              [](bool val) -> parsed::map::key { return val; },
+              [](iobuf& val) -> parsed::map::key { return std::move(val); },
+              [](const auto&) -> parsed::map::key {
+                  throw std::runtime_error(
+                    "invariant: unable to convert type to map key");
+              });
+        }
```

Comment:
I'm kind of surprised we don't catch missing key/value numbers and throw as an invalid input. Is that because fields in general are optional in protobuf?

Reply:
Yes. Everything is optional. in proto3, they really try to steer people away from required because that is hard to ensure compatibility with. Many types (scalars) have defaults if they are "missing".

see: https://protobuf.dev/programming-guides/dos-donts/#add-required

Reply:
Yea I knew about of the great optional reckoning of proto3, but I didn't realize that extended to individual map fields. In this case, it isn't even really about handling missing fields though, or even about optional fields, right? It's that the descriptor is missing a field that it should have?

Though maybe that's what you were getting at by saying maps are weird :)

Reply:
Follow-up question, do message::fields need to be allowed to be monostate as well?

Reply:
> Follow-up question, do message::fields need to be allowed to be monostate as well?

No we just don't write entries in the map for missing fields (it's sparse). The map type is unique because we are using an actual map (for key deduplication), we need to be able to represent a missing value.

Reply:
> In this case, it isn't even really about handling missing fields though, or even about optional fields, right? It's that the descriptor is missing a field that it should have?

The descriptors are value (they have a key and value fields always). So I don't understand. This probably isn't possible output in any proto library but afaik is how to handle missing map entry parts (they just default).

Reply:
Sorry I didn't express that very clearly. I meant that the key and value field numbers seem like special numbers that the value _should_ have by virtue of being a map. But in any case it makes sense to still handle the missing case gracefully.

Reply:
Seems like there are cases where golang will omit a value: https://github.com/golang/protobuf/issues/1312